pkg/infra/pulumi_aws/iac/sanitization/sanitizer.ts

```diff
@@ -0,0 +1,153 @@
+    const violations = new Array<string>();
+    if (options.minLength != null && name.length < options.minLength) {
+        violations.push(`Invalid resource name: "${name}": length < minLength (${options.minLength})`)
+    }
+    if (options.maxLength != null && name.length > options.maxLength) {
+        violations.push(`Invalid resource name: "${name}": length > maxLength (${options.maxLength})`)
+    }
+
+    violations.push(...(options.rules?.filter(r => !r.validate(name)).map(r => `Naming rule violated: ${r.description}`) || []));
+    return violations;
+}
+
+function simpleMerge(resourceName: ResourceName): string {
+    const prefix = resourceName.prefix ? resourceName.prefix : "";
+    const name = resourceName.name ? resourceName.name : "";
+    const suffix = resourceName.suffix ? resourceName.suffix : "";
+    const separator = resourceName.separator ? resourceName.separator : "";
+    return `${prefix}${prefix ? separator : ""}${name}${suffix ? separator : ""}${suffix}`
+}
+
+export interface SanitizationRule extends ValidationRule {
+    fix?: FixFunc
+}
+
+type FixFunc = (string) => string
+
+export interface ValidationRule {
+    description: string
+
+    validate(string): boolean
+}
+
+type NamingStrategy = (resourceName: ResourceName) => string
+type ShorteningStrategy = (name: ResourceName, maxLength: number) => ResourceName
+
+interface ResourceNameOptions extends Partial<SanitizationOptions> {
+    namingStrategy?: NamingStrategy
+    shorteningStrategy?: ShorteningStrategy
+}
+
+
+export function shortenPrefix(resourceName: ResourceName, maxLength: number): ResourceName {
+    const prefix = resourceName.prefix ? resourceName.prefix : ""
+    const name = resourceName.name ? resourceName.name : ""
+    const suffix = resourceName.suffix ? resourceName.suffix : ""
+    const separator = resourceName.separator ? resourceName.separator : ""
+    const separatorCount = prefix && suffix ? 2 : (prefix || suffix ? 1 : 0)
+    const maxPrefixLength = maxLength - name.length - suffix.length - (separator.length * separatorCount)
+    const prefixLength = prefix.length > maxPrefixLength ? maxPrefixLength : prefix.length;
+    return {...resourceName, prefix: prefix?.substring(0, prefixLength)}
+}
+
+export function shortenName(resourceName: ResourceName, maxLength: number): ResourceName {
+    const prefix = resourceName.prefix ? resourceName.prefix : ""
+    const name = resourceName.name ? resourceName.name : ""
+    const suffix = resourceName.suffix ? resourceName.suffix : ""
+    const separator = resourceName.separator ? resourceName.separator : ""
+    const separatorCount = prefix && suffix ? 2 : (prefix || suffix ? 1 : 0)
+    const maxNameLength = maxLength - prefix.length - suffix.length - (separator.length * separatorCount)
+    let nameLength = name.length > maxNameLength ? maxNameLength : name.length;
+    return {...resourceName, name: name.substring(0, nameLength)}
+}
```

Comment:
we will likely need each portion of the name to be shortened so we dont run into conflicts. If someones app name is = allowed length for a resource, all those resources will conflict

Reply:
Yeah, there's definitely room for additional name shortening strategies. (like your example -- _shorten all name components using the supplied separator_).

Reply:
> im wondering if we could simplify it for now since we will want this in go eventually and do something like, given resource name shorten the app name + resource name to max length, then have the regex it needs to match

By "simpliyfy", are you referring to changing the API to support supplying just `name` and then inferring `prefix` and `suffix` where possible based on an optionally supplied separator or simplifying the underlying implementation?